gpii/node_modules/gpii-oauth2/gpii-oauth2-datastore/src/DbDataStoreUtils.js

```diff
@@ -0,0 +1,397 @@
+    gpiiToken: "gpiiToken",
+    client: "client",
+    authDecision: "authDecision",
+    authCode: "authCode",
+    clientCredentialsToken: "clientCredentialsToken"
+};
+
+gpii.oauth2.dbDataStore.findRecord = function (dataSource, directModel, valueNotEmpty, dataProcessFunc) {
+    dataProcessFunc = dataProcessFunc || gpii.oauth2.dbDataStore.cleanUpDoc;
+    var promiseTogo = fluid.promise();
+
+    // Verify required field values are provided instead of undefined
+    var emptyFields = gpii.oauth2.dbDataStore.verifyEmptyFields(directModel, valueNotEmpty);
+
+    if (emptyFields.length > 0) {
+        var error = gpii.oauth2.composeError(gpii.oauth2.errors.missingInput, {fieldName: emptyFields.join(" & ")});
+        promiseTogo.reject(error);
+    } else {
+        var finalDirectModel = $.extend(true, {}, dataSource.options.directModel, directModel);
+        var promise = dataSource.get(finalDirectModel);
+        promise.then(function (data) {
+            // TODO: The line below that converts an empty object to undefined is to work around an issue with using the
+            // kettle notFoundIsEmpty option with fetching couchDB documents by views. The way that notFoundIsEmpty is
+            // implemented in kettle is that, it returns undefined when encountering a 404 response. However, when querying
+            // couchdb by views, the returned value would not be a 404 http status code even when the doc is not found.
+            // The response would still be an object but with an empty "rows" array. An example response is:
+            // { total_rows: 1, offset: 0, rows: [] }
+            // So, the couchDB returned value needs to be further processed using kettle readPayload option, which eventually
+            // returns an empty object. Note that this issue only occurs when querying CouchDB by a view. when querying CouchDB
+            // directly by an id, 404 status is returned.
+            var result = $.isEmptyObject(data) ? undefined : dataProcessFunc(data);
+            if (result !== undefined && result.isError) {
+                promiseTogo.reject(result);
+            } else {
+                promiseTogo.resolve(result);
+            }
+        }, function (error) {
+            promiseTogo.reject(error);
+        });
+    }
+
+    return promiseTogo;
+};
+
+/*
+ * Find elements in the valueNotEmpty array if that element matches a path of the give obj and its value
+ * is undefined.
+ * @obj (Object): The object whose values are checked against undefined.
+ * @valueNotEmpty (String or Array): The path name(s) to be used to locate value on the give obj.
+ * @return: An array of elements in the valueNotEmpty array provided: 1. the element matches a path name on obj;
+ * 2. the corresponding value of that path name is undefined.
+ * For example, gpii.oauth2.dbDataStore.verifyEmptyFields({"a": 1}, ["a", "b"]) returns ["b"].
+ */
+gpii.oauth2.dbDataStore.verifyEmptyFields = function (obj, valueNotEmpty) {
```

Comment:
"Verify" doesn't seem like a good name for this function. It doesn't check something, it looks like it filters on something.